### PR TITLE
+act #17274 make ForkJoinPool asyncMode configurable

### PR DESCRIPTION
**Pull request by He-Pin**
refs https://github.com/akka/akka/issues/17274

### CODE REVIEW

**Comment by ktoso** [comment]
OK TO TEST

File: akka-actor/src/main/scala/akka/dispatch/AbstractDispatcher.scala [hunk 1]
```diff
@@ -54,19 +54,25 @@ private[akka] trait LoadMetrics { self: Executor ⇒
...
-  final val debug = false // Deliberately without type ascription to make it a compile-time constant
+  final val debug = false
+  // Deliberately without type ascription to make it a compile-time constant
   lazy val actors = new Index[MessageDispatcher, ActorRef](16, _ compareTo _)
+
   def printActors(): Unit =
     if (debug) {
       for {
         d ← actors.keys
-        a ← { println(d + " inhabitants: " + d.inhabitants); actors.valueIterator(d) }
```
Comment (ktoso) [inline akka-actor/src/main/scala/akka/dispatch/AbstractDispatcher.scala]: And one less `;` in the akka codebase, good ;-)

File: akka-actor/src/main/scala/akka/dispatch/AbstractDispatcher.scala [hunk 2]
```diff
@@ -84,14 +90,15 @@ private[akka] object MessageDispatcher {
...
-  import MessageDispatcher._
   import AbstractMessageDispatcher.{ inhabitantsOffset, shutdownScheduleOffset }
+  import MessageDispatcher._
   import configurator.prerequisites
 
   val mailboxes = prerequisites.mailboxes
   val eventStream = prerequisites.eventStream
 
-  @volatile private[this] var _inhabitantsDoNotCallMeDirectly: Long = _ // DO NOT TOUCH!
+  @volatile private[this] var _inhabitantsDoNotCallMeDirectly: Long = _
```
Comment (ktoso) [inline akka-actor/src/main/scala/akka/dispatch/AbstractDispatcher.scala]: Would be nice to keep those comments in-line (multiple cases like this one)

**Comment by hepin1989** [comment]
[error] (akka-actor/*:mimaReportBinaryIssues) akka-actor: Binary compatibility check failed!

File: akka-actor/src/main/scala/akka/dispatch/AbstractDispatcher.scala (hunk 1 above)
Comment (2beaucoup) [inline akka-actor/src/main/scala/akka/dispatch/AbstractDispatcher.scala]: I can still see it. ;)

Comment (ktoso) [inline akka-actor/src/main/scala/akka/dispatch/AbstractDispatcher.scala]: lol yeah - wishful thinking from my side

File: akka-actor/src/main/scala/akka/dispatch/AbstractDispatcher.scala [hunk 3]
```diff
@@ -422,9 +441,11 @@ class ForkJoinExecutorConfigurator(config: Config, prerequisites: DispatcherPrer
   }
 
   class ForkJoinExecutorServiceFactory(val threadFactory: ForkJoinPool.ForkJoinWorkerThreadFactory,
-                                       val parallelism: Int) extends ExecutorServiceFactory {
-    def createExecutorService: ExecutorService = new AkkaForkJoinPool(parallelism, threadFactory, MonitorableThreadFactory.doNothing)
+                                       val parallelism: Int,
+                                       val asyncMode: Boolean) extends ExecutorServiceFactory {
```
Comment (ktoso) [inline akka-actor/src/main/scala/akka/dispatch/AbstractDispatcher.scala]: That's the binary compat failure. Technically it's not akka-private so people may have used it.
We could add an additional constructor here to be binary compatible - with the default value as before-this-patch.

// cc @rkuhn

File: akka-actor/src/main/scala/akka/dispatch/AbstractDispatcher.scala (hunk 1 above)
Comment (2beaucoup) [inline akka-actor/src/main/scala/akka/dispatch/AbstractDispatcher.scala]: +1

Comment (hepin1989) [inline akka-actor/src/main/scala/akka/dispatch/AbstractDispatcher.scala]: @2beaucoup you can't see it now:)

Comment (2beaucoup) [inline akka-actor/src/main/scala/akka/dispatch/AbstractDispatcher.scala]: Yeah. But all the whitespace changes make reviewing hard. :(

Comment (hepin1989) [inline akka-actor/src/main/scala/akka/dispatch/AbstractDispatcher.scala]: sorry for that changes :(``

Comment (rkuhn) [inline akka-actor/src/main/scala/akka/dispatch/AbstractDispatcher.scala]: You know that you can add a query parameter '?w=1' to the diff view to suppress whitespace changes, right? The only downside is that for comments you'll have to use a pristine tab.

File: akka-actor/src/main/scala/akka/dispatch/AbstractDispatcher.scala [hunk 4]
```diff
@@ -60,13 +60,18 @@ private[akka] object MessageDispatcher {
 
   // dispatcher debugging helper using println (see below)
   // since this is a compile-time constant, scalac will elide code behind if (MessageDispatcher.debug) (RK checked with 2.9.1)
-  final val debug = false // Deliberately without type ascription to make it a compile-time constant
+  final val debug = false
+  // Deliberately without type ascription to make it a compile-time constant
```
Comment (patriknw) [inline akka-actor/src/main/scala/akka/dispatch/AbstractDispatcher.scala]: move the comment above the `val` if not on same line

File: akka-actor/src/main/scala/akka/dispatch/AbstractDispatcher.scala (hunk 1 above)
Comment (2beaucoup) [inline akka-actor/src/main/scala/akka/dispatch/AbstractDispatcher.scala]: Yeah, I'm aware of this workaround. :)

File: akka-actor/src/main/resources/reference.conf [hunk 5]
```diff
@@ -276,6 +276,9 @@ akka {
 
         # Max number of threads to cap factor-based parallelism number to
         parallelism-max = 64
+
+        # Enable establishing local first-in-first-out scheduling mode for forked tasks that are never joined.
+        async-mode = on
```
Comment (patriknw) [inline akka-actor/src/main/resources/reference.conf]: default is `false`, so we might want to keep this default `off` for 2.3 to not change behavior?

Comment (hepin1989) [inline akka-actor/src/main/resources/reference.conf]: default is true

Comment (drewhk) [inline akka-actor/src/main/resources/reference.conf]: The comment is a bit confusing. Akka users might have not used FJP therefore "forked tasks that are never joined" does not help. Also the tasks that Akka schedules are never joined anyway.

**Comment by drewhk** [comment]
LGTM after binary compat resolved, and a better comment in the reference.conf

File: akka-actor/src/main/resources/reference.conf (hunk 5 above)
Comment (hepin1989) [inline akka-actor/src/main/resources/reference.conf]: so just FIFO or LIFO?I don't how to describe that in a proper way.could you please just comment here so I could copy and paste?

Comment (patriknw) [inline akka-actor/src/main/resources/reference.conf]: I see, akka currently use `true`. Good

File: akka-actor/src/main/resources/reference.conf [hunk 6]
```diff
@@ -276,6 +276,10 @@ akka {
 
         # Max number of threads to cap factor-based parallelism number to
         parallelism-max = 64
+
+        # Set to first-in-first-out scheduling mode when you set "async-mode = on" or
+        # last-in-first-out scheduling mode when you set "async-mode = off".
+        async-mode = on
```
Comment (rkuhn) [inline akka-actor/src/main/resources/reference.conf]: I know that this comes from FJP, but it still reads awkwardly: “asynchrony” cannot really be switched on and off with this toggle. I’d prefer to be explicit and parse the strings `LIFO` and `FIFO` (explaining that one is a stack and the other a queue should be enough to give people an idea of what this does).

**Comment by rkuhn** [comment]
Thanks @hepin1989!

**Comment by patriknw** [comment]
LGTM

File: akka-actor/src/main/scala/akka/dispatch/AbstractDispatcher.scala [hunk 7]
```diff
@@ -432,12 +455,18 @@ class ForkJoinExecutorConfigurator(config: Config, prerequisites: DispatcherPrer
         m.withName(m.name + "-" + id)
       case other ⇒ other
     }
+    val asyncMode = config.getString("task-peeking-mode") match {
+      case "FIFO"      ⇒ true
+      case "LIFO"      ⇒ false
+      case unsupported ⇒ throw new IllegalStateException(s""""task-peeking-mode" could only set to "FIFO" or "LILO",but you provided:[$unsupported].""")
```
Comment (2beaucoup) [inline akka-actor/src/main/scala/akka/dispatch/AbstractDispatcher.scala]: this msg can be improved and lacks some whitespaces. ;)

Comment (patriknw) [inline akka-actor/src/main/scala/akka/dispatch/AbstractDispatcher.scala]: and it should be `IllegalArgumentException`

**Comment by drewhk** [comment]
Do we want to have this in 2.3.11? This will also need a forward port to master.

**Comment by patriknw** [comment]
It's ready and I vote fore including it in 2.3.11.

**Comment by rkuhn** [comment]
Agreed.

**Comment by hepin1989** [comment]
will rebase

**Comment by patriknw** [comment]
why closed? you don't have to open a new PR for rebasing

**Comment by hepin1989** [comment]
sorry,I force updated on another computer,but the code is not right,and when I merge the conflict,seems like I missed something.so I came to here.

**Comment by patriknw** [comment]
ok, superseded by https://github.com/akka/akka/pull/17423

**Pull request closed without merging**